src/main/java/org/jenkinsci/remoting/util/VersionNumber.java

```diff
@@ -0,0 +1,579 @@
+                    break;
+                }
+            }
+        }
+
+        public int compare(Item item) {
+            if (item == null) {
+                if (size() == 0) {
+                    return 0; // 1-0 = 1- (normalize) = 1
+                }
+                Item first = (Item) get(0);
+                return first.compare(null);
+            }
+
+            switch (item.getType()) {
+                case INTEGER_ITEM:
+                    return -1; // 1-1 < 1.0.x
+
+                case STRING_ITEM:
+                    return 1; // 1-1 > 1-sp
+
+                case LIST_ITEM:
+                    if (item instanceof ListItem) {
+                        Iterator left = iterator();
+                        Iterator right = ((ListItem) item).iterator();
+
+                        while (left.hasNext() || right.hasNext()) {
+                            Item l = left.hasNext() ? (Item) left.next() : null;
+                            Item r = right.hasNext() ? (Item) right.next() : null;
+
+                            // if this is shorter, then invert the compare and mul with -1
+                            int result;
+                            if (l == null) {
+                                if (r == null) {
+                                    result = 0;
+                                } else {
+                                    result = -1 * r.compare(null);
+                                }
+                            } else {
+                                result = l.compare(r);
+                            }
+
+                            if (result != 0) {
+                                return result;
+                            }
+                        }
+
+                        return 0;
+                    }
+
+                case WILDCARD_ITEM:
+                    return -1;
+
+                default:
+                    throw new RuntimeException("invalid item: " + item.getClass());
+            }
+        }
+
+        public String toString() {
+            StringBuilder buffer = new StringBuilder("(");
+            for (Iterator<Item> iter = iterator(); iter.hasNext(); ) {
+                buffer.append(iter.next());
+                if (iter.hasNext()) {
+                    buffer.append(',');
+                }
+            }
+            buffer.append(')');
+            return buffer.toString();
+        }
+    }
+
+    public VersionNumber(String version) {
+        parseVersion(version);
+    }
+
+    private void parseVersion(String version) {
+        this.value = version;
+
+
+        items = new ListItem();
+
+        Matcher matcher = SNAPSHOT.matcher(version);
+        if (matcher.matches()) {
+            snapshot = matcher.group(1);
+            version = version.substring(0, matcher.start(1)) + "-SNAPSHOT";
+        }
+        version = version.toLowerCase(Locale.ENGLISH);
+
+        ListItem list = items;
+
+        Stack<Item> stack = new Stack<Item>();
```

Comment:
Better to use

```suggestion
        Deque<Item> stack = new ArrayDeque<>();
```

And then you can still push() and pop() as a stack. Stack is deprecated in Java.